### PR TITLE
Jormungandr: Journey filters optimization

### PR DESCRIPTION
**Pull request by xlqian**
https://jira.kisio.org/browse/NAVP-861

I used the following request, with which kraken is called  99 times (i requested 100 journeys) for the benchmark
```bash
time curl 'http://localhost:5000/v1/coverage/stif/journeys?from=stop_area%3A0%3ASA%3A8768600&to=stop_area%3A0%3ASA%3A8775860&datetime=20180423T083802&count=100' > /dev/null
```
In this [loop:](https://github.com/CanalTP/navitia/pull/2404/files#diff-f050c03e2b24acf6b927fade7d249a87L813):

|   time used in   | C++(Kraken) | Python |
|--------|--------|--------|
| Before |    4.46s    | 3.44s        | 
| After  |      4.41s  |  0.13s      | 

another request, with which kraken is called for 30 times (100 journeys are requested, same as before)
```bash
time curl 'http://localhost:5000/v1/coverage/stif/journeys?from=2.284794%3B48.911762&to=2.37715%3B48.846781&count=100&datetime=20180423T083802&' > /dev/null
```

|   time used in     | C++(Kraken) | Python |
|--------|--------|--------|
| Before |   3.21s   | 1.2s        | 
| After  |     3.30s  |  0.15s      |

### CODE REVIEW

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 1]
```diff
@@ -207,10 +199,10 @@ def _filter_too_short_heavy_journeys(journeys, request):
     Typically you don't take your car for only 2 minutes
     """
     logger = logging.getLogger(__name__)
+
     for journey in journeys:
-        if to_be_deleted(journey):
-            continue
         on_bss = False
+        is_dead = False
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: I hate this pattern, better to do that in a function, no?

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 2]
```diff
@@ -308,35 +312,43 @@ def _filter_min_transfers(journeys, instance, request):
...
+        elif dp == 'only' and 'non_pt' not in j.tags:
             logger.debug("the journey {} uses pt, we delete it as param direct_path=only"
                          .format(j.internal_id))
-            mark_as_dead(j, "direct_path_only")
-
+            mark_as_dead(j, request, "direct_path_only")
+            is_dead = True
+        if not is_dead:
+            yield j
+        elif request.get('debug'):
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: That's not done in the others, no?

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 3]
```diff
@@ -810,30 +810,47 @@ def fill_journeys(self, request_type, api_request, instance):
...
-            responses.extend(tmp_resp)  # we keep the error for building the response
-            if nb_journeys(tmp_resp) == 0:
+            new_resp = self.call_kraken(request_type, request, instance, krakens_call)
+            _tag_by_mode(new_resp)
+            _tag_direct_path(new_resp)
+            _tag_bike_in_pt(new_resp)
+            journey_filter._filter_too_long_journeys(new_resp, request)
+            if nb_journeys(new_resp) == 0:
                 # no new journeys found, we stop
+                responses.extend(new_resp)
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: Maybe add a comment saying that we add the to_delete tagged journeys

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 4]
```diff
@@ -810,30 +810,47 @@ def fill_journeys(self, request_type, api_request, instance):
...
                 break
 
-            request = self.create_next_kraken_request(request, tmp_resp)
+            request = self.create_next_kraken_request(request, new_resp)
 
             # we filter unwanted journeys by side effects
-            journey_filter.filter_journeys(responses, instance, api_request)
+            filtered_new_resp = journey_filter.filter_journeys(new_resp, instance, api_request)
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: The comment seems unsync

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 5]
```diff
@@ -810,30 +810,47 @@ def fill_journeys(self, request_type, api_request, instance):
...
+
+            #duplicate the iterator
+            tmp1, tmp2 = itertools.tee(filtered_new_resp)
+            qualified_journeys = journey_filter.get_qualified_journeys(responses)
+
+            journeys_pool = itertools.chain(
+                itertools.combinations(tmp1, 2),
+                itertools.product(tmp2, qualified_journeys),
+            )
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: A comment on that might be great.

File: source/jormungandr/jormungandr/scenarios/qualifier.py [hunk 6]
```diff
@@ -184,12 +184,7 @@ def choose_standard(journeys, best_criteria):
 
 #comparison of 2 fields. 0=>equality, 1=>1 better than 2
 def compare_minus(field_1, field_2):
-    if field_1 == field_2:
-        return 0
-    elif field_1 < field_2:
-        return 1
-    else:
-        return -1
+    return cmp(field_2, field_1)
```
Comment (TeXitoi) [inline source/jormungandr/jormungandr/scenarios/qualifier.py]: :+1:

**Comment by xlqian** [comment]
All Tests Artemis have passed: https://ci.navitia.io/view/artemis/job/Artemis/1708/console

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 7]
```diff
@@ -83,15 +73,16 @@ def final_filter_journeys(response_list, instance, request):
...
-    journeys = [j for r in response_list for j in r.journeys]
 
     final_line_filter = get_or_default(request, '_final_line_filter', False)
     if final_line_filter:
-        _filter_similar_line_journeys(journeys, request)
+        journeys = get_qualified_journeys(response_list)
+        journeys_pool = itertools.combinations(journeys, 2)
+        _filter_similar_line_journeys(journeys_pool, request)
 
+    journeys = get_qualified_journeys(response_list)
```
Comment (mbounabi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:104]: :+1:

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 8]
```diff
@@ -810,30 +810,47 @@ def fill_journeys(self, request_type, api_request, instance):
...
-            _tag_direct_path(tmp_resp)
-            _tag_bike_in_pt(tmp_resp)
-            journey_filter._filter_too_long_journeys(tmp_resp, request)
-            responses.extend(tmp_resp)  # we keep the error for building the response
-            if nb_journeys(tmp_resp) == 0:
+            new_resp = self.call_kraken(request_type, request, instance, krakens_call)
+            _tag_by_mode(new_resp)
+            _tag_direct_path(new_resp)
+            _tag_bike_in_pt(new_resp)
+            journey_filter._filter_too_long_journeys(new_resp, request)
```
Comment (mbounabi) [inline source/jormungandr/jormungandr/scenarios/new_default.py]: The aim of the function '_filter_too_long_journeys' is to filter journeys that starts after a period (pseudo_duration) that is later in the future.
Better to call it 'too_late_journeys'? 'too_long journeys' seems that the duration of journey is too long

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 9]
```diff
@@ -47,34 +49,22 @@ def delete_journeys(responses, request):
...
-    _filter_direct_path(journeys, instance, request)
-
-    return response_list
+    if request['debug']:
+        [_debug_journey(j) for j in get_qualified_journeys(responses)]
 
+    return compose(functools.partial(_filter_too_short_heavy_journeys, request=request),
+                   functools.partial(_filter_too_long_waiting, request=request),
+                   functools.partial(_filter_max_successive_physical_mode, instance=instance, request=request),
+                   functools.partial(_filter_min_transfers, instance=instance, request=request),
```
Comment (mbounabi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: 'instance' not used in the function

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 10]
```diff
@@ -47,34 +49,22 @@ def delete_journeys(responses, request):
...
+                   functools.partial(_filter_direct_path, instance=instance, request=request),
```
Comment (mbounabi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: 'instance' not used in the function

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 11]
```diff
@@ -222,33 +214,42 @@ def _filter_too_short_heavy_journeys(journeys, request):
...
                 continue
             logger.debug("the journey {} has a too long waiting, we delete it".format(j.internal_id))
-            mark_as_dead(j, "too_long_waiting")
+            mark_as_dead(j, request, "too_long_waiting")
+            is_dead = True
             break
-
+        if not is_dead:
+            yield j
+        elif request.get('debug'):
```
Comment (mbounabi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: Get the boolean 'debug' before loop or even before the call to the function to get it just once

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 12]
```diff
@@ -155,21 +146,21 @@ def to_be_deleted(journey):
     return 'to_delete' in journey.tags
 
 
-def mark_as_dead(journey, *reasons):
+def mark_as_dead(journey, request, *reasons):
```
Comment (mbounabi) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: Get the boolean 'debug' before calling the function to avoid passing the whole request object

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 13]
```diff
@@ -47,33 +49,43 @@ def delete_journeys(responses, request):
...
 
-    _filter_similar_vj_journeys(journeys, request)
-
-    _filter_too_long_waiting(journeys, request)
-
-    _filter_max_successive_physical_mode(journeys, instance, request)
+    min_nb_transfers = get_or_default(request, 'min_nb_transfers', 0)
 
-    _filter_min_transfers(journeys, instance, request)
+    # Note that we use the functools.partial to capture the arguments
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: for python lovers: 
"I want to capture arguments, use lambda or partial?" 
TL;DR
https://stackoverflow.com/q/3252228/1614576
https://stackoverflow.com/a/24653777/1614576

File: source/jormungandr/jormungandr/utils.py [hunk 14]
```diff
@@ -563,3 +566,85 @@ def worker():
...
+
+    >>> f = lambda l: (x for x in l if x%3 == 0)
+    >>> g = lambda l: (x for x in l if x%5 == 0)
+    >>> c = compose(f, g)
+    >>> list(c(range(45)))
+    [0, 15, 30]
+    """
+    return lambda obj: functools.reduce(lambda prev, f: f(prev), funs, obj)
+
+class ComposedFilter(object):
```
Comment (benoit-bst) [inline source/jormungandr/jormungandr/utils.py:602]: :heart:

File: source/jormungandr/jormungandr/scenarios/qualifier.py (hunk 6 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/qualifier.py]: my bad, `cmp` is removed from python3, we should use `(a > b) - (a < b)` https://docs.python.org/3.4/whatsnew/3.0.html#ordering-comparisons

**Review by benoit-bst — APPROVED** [review]

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 15]
```diff
@@ -200,84 +214,82 @@ def exceed_min_duration(current_section, journey, request,  min_duration):
...
     Heavy fallback mode are Bike and Car, bss is not considered as one.
     Typically you don't take your car for only 2 minutes
     """
     logger = logging.getLogger(__name__)
-    for journey in journeys:
-        if to_be_deleted(journey):
+
+    on_bss = False
+    is_debug = request.get('debug', False)
+    for s in journey.sections:
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: Since we now have the duration for each mode, we can do an early return.
```
bike_duration==0: true
car_duration==0: true
car_duration<_min_car: false
bike_duration<_min_bike: false (only if bike and bss are distinct in Durations)
```
we will only have to scan the whole journey when there is more car/bike than the min to check if one section isn't bellow this minimum

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 16]
```diff
@@ -200,84 +214,82 @@ def exceed_min_duration(current_section, journey, request,  min_duration):
...
+
+def filter_too_long_waiting(journey, is_debug):
     """
     filter journeys with a too long section of type waiting
     """
     logger = logging.getLogger(__name__)
-    for j in journeys:
-        if to_be_deleted(j):
+
+    for s in journey.sections:
```
Comment (kinnou02) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:277]: Same trick here, if the duration of the journey is less than max_duration there is no need to check each sections. 
Also I think we have the number of transfers of a journey, if there is no transfers it won't have any waiting sections

**Review by kinnou02 — APPROVED** [review]
:+1:

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 13 above)
Comment (nonifier) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: would it be a little more readable to create the partials on a separate line. Something along those lines:
```python
partials = [
   partial(filter_too_short_heavy_journeys, request),
   partial(filter_too_long_waiting, is_debug=is_debug),
   partial(filter_min_transfers, is_debug=is_debug, min_nb_transfers=min_nb_transfers)
]
composed_filter = ComposedFilter()
for p in partials: 
   composed_filter.add_filter(p);
```

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 17]
```diff
@@ -407,34 +412,35 @@ def way_later(request, journey1, journey2):
...
-def _filter_too_long_journeys(responses, request):
+def _filter_too_late_journeys(responses, request):
     """
     Filter not coherent journeys
 
     The aim is to keep that as simple as possible
     """
-    # for clarity purpose we build a temporary list
-    journeys = [j for r in responses for j in r.journeys if 'non_pt' not in j.tags]
+    # for clarity purpose we build a temporary iterator
```
Comment (nonifier) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: Is this not an iterator but rather a generator ? :wink:

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 18]
```diff
@@ -522,3 +528,13 @@ def _datetime_to_str(ts):
         duration=datetime.timedelta(seconds=journey.duration),
         fallback=datetime.timedelta(seconds=fallback_duration(journey)),
         sec=" - ".join(sections)))
+
+
+def get_qualified_journeys(responses):
+    """
+
+    :param responses: protobuf
+    :return: iterator of journeys
```
Comment (nonifier) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py]: `iterator` -> `iterable` of journeys :D

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 19]
```diff
@@ -810,30 +811,64 @@ def fill_journeys(self, request_type, api_request, instance):
...
-            if nb_journeys(tmp_resp) == 0:
+            new_resp = self.call_kraken(request_type, request, instance, krakens_call)
+            _tag_by_mode(new_resp)
+            _tag_direct_path(new_resp)
+            _tag_bike_in_pt(new_resp)
+            journey_filter._filter_too_late_journeys(new_resp, request)
+            if nb_journeys(new_resp) == 0:
                 # no new journeys found, we stop
+                # we still append the new_resp because there are journeys that a tagged as dead probably
+                responses.extend(new_resp)
```
Comment (nonifier) [inline source/jormungandr/jormungandr/scenarios/new_default.py:829]: :+1:

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 20]
```diff
@@ -810,30 +811,64 @@ def fill_journeys(self, request_type, api_request, instance):
...
+                # First step: compare journeys from the new response only
+                itertools.combinations(tmp1, 2),
+                # Second step:
+                # we use the itertools.product to create combinations between qualified journeys and new journeys
+                # Ex:
+                # new_journeys = [n_1, n_2]
+                # qualified_journeys = [q_1, q_2, q_3]
+                # itertools.product(new_journeys, qualified_journeys) gives combinations as follows:
+                # (n_1, q_1), (n_1, q_2),(n_1, q_3),(n_2, q_1),(n_2, q_2),(n_2, q_3)
+                itertools.product(tmp2, qualified_journeys),
```
Comment (nonifier) [inline source/jormungandr/jormungandr/scenarios/new_default.py:858]: :confused: 
I'm not sure what's really going on here with this business logic, but when I see way more comments than actual code, my guts are telling me that something is fishy here.

File: source/jormungandr/jormungandr/scenarios/qualifier.py [hunk 21]
```diff
@@ -184,12 +184,9 @@ def choose_standard(journeys, best_criteria):
...
 def compare_minus(field_1, field_2):
-    if field_1 == field_2:
-        return 0
-    elif field_1 < field_2:
-        return 1
-    else:
-        return -1
+    # ordering-comparisons
+    # https://docs.python.org/3.4/whatsnew/3.0.html#ordering-comparisons
+    return (field_2 > field_1) - (field_2 < field_1)
```
Comment (nonifier) [inline source/jormungandr/jormungandr/scenarios/qualifier.py:189]: :open_mouth: Sometimes, python looks amazing and sometimes is not all the time....

File: source/jormungandr/jormungandr/utils.py [hunk 22]
```diff
@@ -563,3 +566,85 @@ def worker():
...
+    [0, 30]
+    """
+    def __init__(self):
+        self.filters = []
+
+    def add_filter(self, pred):
+        self.filters.append(lambda iterable: (i for i in iterable if pred(i)))
+        return self
+
+    def compile(self):
```
Comment (nonifier) [inline source/jormungandr/jormungandr/utils.py]: I would rather see `def compose_filters` or `apply_filters` to be more consistent.

**Review by nonifier — CHANGES_REQUESTED** [review]

File: source/jormungandr/jormungandr/scenarios/new_default.py (hunk 20 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/new_default.py:858]: @nonifier  :D most of them are parts of the example

File: source/jormungandr/jormungandr/scenarios/qualifier.py (hunk 21 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/qualifier.py:189]: yep, totally wtf

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 15 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py]: LOL I understand your point and I tried really hard to optimize this, my final codes don't appear more compact than the original one :rofl: we can discuss about this later

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 16 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py:277]: OK for this one

**Review by nonifier — APPROVED** [review]
:+1:

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 23]
```diff
@@ -188,96 +201,117 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
...
+    # There are no bike/bss/car fallback
+    if journey.durations.bike == journey.durations.car == 0:
+        return True
+
+    # Early return 2
+    # Durations of every mode are >= min
+    if journey.durations.bike >= request.get('_min_bike', float('inf')) and \
+            journey.durations.car >= request.get('_min_car', float('inf')):
+        return True
+
```
Comment (xlqian) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:234]: @kinnou02 is this ok for you?

**Comment by nonifier** [comment]
@xlqian : Is this ready to be merged ?

**Comment by xlqian** [comment]
@nonifier Nein

**Review by mbounabi — APPROVED** [review]

File: source/jormungandr/jormungandr/utils.py (hunk 14 above)
Comment (pbougue) [inline source/jormungandr/jormungandr/utils.py:602]: How about the order in which filters are applied? This is really important (to apply quicker filters first), should be documented and maybe tested if possible (already tested on compose(), harder on ComposeFilter)

But definitelly :heart_eyes: for this code :)

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 24]
```diff
@@ -284,11 +285,7 @@ def test_journeys_equality_test_same_journeys():
 
 
 def journeys_gen(list_responses):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:287]: The naming is not great : `journey_pairs_gen` would be clearer
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 25]
```diff
@@ -155,21 +167,21 @@ def to_be_deleted(journey):
...
+def filter_similar_vj_journeys(journeys, request):
     _filter_similar_journeys(journeys, request, similar_journeys_vj_generator)
 
 
 def _filter_similar_line_journeys(journeys, request):
     _filter_similar_journeys(journeys, request, similar_journeys_line_generator)
 
 
-def _filter_similar_journeys(journeys, request, similar_journey_generator):
+def _filter_similar_journeys(journeys_pool, request, similar_journey_generator):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:184]: I'd name the parameter `journey_pairs_pool` (same on pretty much all filters I guess)
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 26]
```diff
@@ -948,7 +944,7 @@ def test_heavy_journey_bss():
     journey.sections[-1].street_network.mode = response_pb2.Walking
     journey.sections[-1].duration = 5
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:945]: missing an update of `durations` to really test something

File: source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py [hunk 27]
```diff
@@ -984,70 +980,80 @@ def test_activate_deactivate_min_bike():
     journey.sections[-1].street_network.mode = response_pb2.Bike
     journey.sections[-1].duration = 7
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/tests/journey_compare_tests.py:981]: missing `durations` update to really test something

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 28]
```diff
@@ -810,30 +811,64 @@ def fill_journeys(self, request_type, api_request, instance):
 
         responses = []
         nb_try = 0
+        nb_qualified_journeys = nb_journeys(responses)
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/new_default.py:814]: We should comment to highlight that it's initialized at 0 and it's normal.

File: source/jormungandr/jormungandr/scenarios/new_default.py [hunk 29]
```diff
@@ -810,30 +811,64 @@ def fill_journeys(self, request_type, api_request, instance):
...
+            tmp1, tmp2 = itertools.tee(filtered_new_resp)
+            qualified_journeys = journey_filter.get_qualified_journeys(responses)
+
+            # now we want to filter similar journeys in the new response which is done in 2 steps
+            # In the first step, we compare journeys from the new response only , 2 by 2
+            # hopefully, it may lead to some early return for the second step to improve the perf a little
+            # In the second step, we compare the journeys from the new response with those that have been qualified
+            # already in the former iterations
+            # note that the journeys_pool is a list of 2-element tuple of journeys
+            journeys_pool = itertools.chain(
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/new_default.py:848]: to be renamed `journey_pairs_pool`

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 30]
```diff
@@ -47,33 +49,42 @@ def delete_journeys(responses, request):
...
-    journeys = [j for r in response_list for j in r.journeys]
-
+    is_debug = request.get('debug')
     #DEBUG
-    for j in journeys:
-        _debug_journey(j)
-
-    _filter_too_short_heavy_journeys(journeys, request)
+    if is_debug:
+        [_debug_journey(j) for j in get_qualified_journeys(responses)]
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:60]: Why not on all the journeys?

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 31]
```diff
@@ -47,33 +49,42 @@ def delete_journeys(responses, request):
...
 
-    _filter_similar_vj_journeys(journeys, request)
+    min_nb_transfers = get_or_default(request, 'min_nb_transfers', 0)
 
-    _filter_too_long_waiting(journeys, request)
+    # Note that we use the functools.partial to capture the arguments
+    filters = [partial(filter_too_short_heavy_journeys, request=request),
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:65]: Probably nicer to extract usefull params from request before, like you do on other filters
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 32]
```diff
@@ -47,33 +49,42 @@ def delete_journeys(responses, request):
...
+               partial(filter_too_long_waiting, is_debug=is_debug),
+               partial(filter_min_transfers, is_debug=is_debug, min_nb_transfers=min_nb_transfers)]
 
-    _filter_max_successive_physical_mode(journeys, instance, request)
+    # we add more filters in some special cases:
+    max_successive = get_or_default(request, '_max_successive_physical_mode', 0)
+    if max_successive != 0:
+        limit_id = instance.successive_physical_mode_to_limit_id
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:72]: to be renamed `limited_mode_id = ... # typically : physical_mode:bus`

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 33]
```diff
@@ -188,96 +201,117 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
...
 
-        if bus_count > max_successive_physical_mode:
-            logger.debug("the journey {} has a too much successive {}, we delete it".
-                format(j.internal_id, instance.successive_physical_mode_to_limit_id))
-            mark_as_dead(j, "too_much_successive_physical_mode")
+    if bus_count > max_successive_physical_mode:
+        logger.debug("the journey {} has a too much successive {}, we delete it".
+            format(journey.internal_id, successive_physical_mode_to_limit_id))
+        mark_as_dead(journey, is_debug, "too_much_successive_physical_mode")
+        return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:312]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 34]
```diff
@@ -188,96 +201,117 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
...
-                break
-
-
-def _filter_too_long_waiting(journeys, request):
+
+        if s.street_network.mode == response_pb2.Car \
+                and exceed_min_duration(s, journey, request, request['_min_car']):
+            logger.debug("the journey {} has not enough car, we delete it".format(journey.internal_id))
+            mark_as_dead(journey, is_debug, "not_enough_car")
+            return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:250]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 35]
```diff
@@ -188,96 +201,117 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
...
+        if not on_bss \
+                and s.street_network.mode == response_pb2.Bike \
+                and exceed_min_duration(s, journey, request, request['_min_bike']):
+            logger.debug("the journey {} has not enough bike, we delete it".format(journey.internal_id))
+            mark_as_dead(journey, is_debug, "not_enough_bike")
+            return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:256]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py (hunk 23 above)
Comment (pbougue) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py:234]: On average, we are probably loosing time with this early return
> :heavy_check_mark: removed in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 36]
```diff
@@ -188,96 +201,117 @@ def _filter_similar_journeys(journeys, request, similar_journey_generator):
...
-            if s.duration < 4 * 60 * 60:
-                continue
-            logger.debug("the journey {} has a too long waiting, we delete it".format(j.internal_id))
-            mark_as_dead(j, "too_long_waiting")
-            break
+        if s.duration < max_waiting:
+            continue
+        logger.debug("the journey {} has a too long waiting, we delete it".format(journey.internal_id))
+        mark_as_dead(journey, is_debug, "too_long_waiting")
+        return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:284]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 37]
```diff
@@ -47,33 +49,42 @@ def delete_journeys(responses, request):
...
 
-    _filter_direct_path(journeys, instance, request)
+    composed_filter = ComposedFilter()
+    [composed_filter.add_filter(f) for f in filters]
 
-    return response_list
+    return composed_filter.compose_filters()(get_qualified_journeys(responses))
 
 
 def final_filter_journeys(response_list, instance, request):
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:90]: the debug mode is not great here along all final_filter_journeys()

Comment (pbougue) [reply source/jormungandr/jormungandr/scenarios/journey_filter.py:90]: Maybe also comment that it's applied only once (no cubic problem).
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 38]
```diff
@@ -407,34 +434,35 @@ def way_later(request, journey1, journey2):
...
+def _filter_too_late_journeys(responses, request):
     """
     Filter not coherent journeys
 
     The aim is to keep that as simple as possible
     """
-    # for clarity purpose we build a temporary list
-    journeys = [j for r in responses for j in r.journeys if 'non_pt' not in j.tags]
+    # for clarity purpose we build a temporary generator
+    journeys = (j for r in responses for j in r.journeys if 'non_pt' not in j.tags
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:444]: Doesn't it depend if we are in debug mode?

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 39]
```diff
@@ -286,57 +320,50 @@ def _filter_too_much_connections(journeys, instance, request):
...
-    for j in journeys:
-        if to_be_deleted(j):
-            continue
-        if get_nb_connections(j) < min_nb_transfers:
-            logger.debug("the journey {} has not enough connections, we delete it".format(j.internal_id))
-            mark_as_dead(j, "not_enough_connections")
+    if get_nb_connections(journey) < min_nb_transfers:
+        logger.debug("the journey {} has not enough connections, we delete it".format(journey.internal_id))
+        mark_as_dead(j, is_debug, "not_enough_connections")
+        return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:344]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 40]
```diff
@@ -286,57 +320,50 @@ def _filter_too_much_connections(journeys, instance, request):
...
     logger = logging.getLogger(__name__)
-    dp = get_or_default(request, 'direct_path', 'indifferent')
 
-    if dp == 'indifferent':
-        return
+    if dp == 'none' and 'non_pt' in journey.tags:
+        logger.debug("the journey {} is direct, we delete it as param direct_path=none"
+                     .format(journey.internal_id))
+        mark_as_dead(journey, is_debug, "direct_path_none")
+        return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:358]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 41]
```diff
@@ -286,57 +320,50 @@ def _filter_too_much_connections(journeys, instance, request):
...
-            mark_as_dead(j, "direct_path_none")
-        if dp == 'only' and 'non_pt' not in j.tags:
-            logger.debug("the journey {} uses pt, we delete it as param direct_path=only"
-                         .format(j.internal_id))
-            mark_as_dead(j, "direct_path_only")
+    elif dp == 'only' and 'non_pt' not in journey.tags:
+        logger.debug("the journey {} uses pt, we delete it as param direct_path=only"
+                     .format(journey.internal_id))
+        mark_as_dead(journey, is_debug, "direct_path_only")
+        return False or is_debug
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:364]: `return is_debug` is the same
> :heavy_check_mark: Done in a following PR

File: source/jormungandr/jormungandr/scenarios/journey_filter.py [hunk 42]
```diff
@@ -286,57 +320,50 @@ def _filter_too_much_connections(journeys, instance, request):
     """
     logger = logging.getLogger(__name__)
     max_additional_connections = get_or_default(request, '_max_additional_connections',
-                                            instance.max_additional_connections)
-
-    min_connections = get_min_connections(journeys)
+                                                instance.max_additional_connections)
+    import itertools
+    it1, it2 = itertools.tee(journeys, 2)
```
Comment (pbougue) [inline source/jormungandr/jormungandr/scenarios/journey_filter.py:325]: Why that change?
If we want to improve perfs a bit on that filter, we should probably sort.

**Review by pbougue — COMMENTED** [review]
Debug mode is not perfectly handled everywhere.
Tests need little corrections.
We could do a more unified filter interface, adding one by one... But it has an important development cost.

File: source/jormungandr/jormungandr/utils.py (hunk 14 above)
Comment (xlqian) [reply source/jormungandr/jormungandr/utils.py:602]: as discussed, the order of filter is respected and yes, this need to be better documented
